### PR TITLE
third_party: update luarocks to 3.x version

### PR DESCRIPTION
**Pull request by zwirec**
Update luarocks to 3.x version and make tarantool compatible with this changes.

This PR also adds new command for `tarantoolctl`: `tarantoolctl rocks test`. This command was added in LuaRocks 3.0.0 and it allows users to test their applications natively including existing rock modules.

See also updated luarocks repo: https://github.com/tarantool/luarocks/tree/luarocks-3.x

### CODE REVIEW

File: extra/dist/tarantoolctl.in [hunk 1]
```diff
@@ -1186,7 +1186,7 @@ local commands = setmetatable({
     }, rocks = {
         func = exit_wrapper(rocks), process = process_remote, help = {
             header =
-                "%s rocks [install|remove|show|search|list]",
+                "%s rocks [install|make|remove|show|search|list|pack|test]",
```
Comment (rosik) [inline extra/dist/tarantoolctl.in]: doc, pack, unpack

**Comment by rosik** [comment]
> Firstly, we should decide how we'll update luarocks repository. I propose the following way: we'll create `tarantool-2` branch based on some luarocks-3 version. Then we'll update the branch with our patchset that we currently maintain upward luarocks-2. Please, use `git cherry-pick -x` to catch commits to allow us understand in the future what was the origin of a change. I think that the current patchset can be cherry-picked to the new branch w/o extra review iteration or maybe with all-in-one PR to briefly look into them as once (I mean a PR into `tarantool-2` branch). If additional patches are needed, then open a separate PR against this branch.
> 
> Now I see two commits that update the submodule. Also there are changes to commands description that were forgotten when some commands were added: AFAIU it should adds help entries for `tarantoolctl rocks <subcommand> --help`. However it seems to be broken now (the command shows the same output for any subcommand). Anyway, these changes also spread across two commits and mixed with luarocks-3 support changes. I propose to do the following.
> 
>     * Factor out all changes that are not related to luarocks-3 support into a separate PR. Add forgotten commands descriptions, check whether `--help` works, fix arguments descriptions — all that here. Those changes are easy and can pass a review faster then remaining ones.
> 
>     * Update the submodule (to a temporary branch for now, say, existing `luarocks-3.x`; then we'll change it to carefully prepared `tarantool-2`). Make changes that are necessary for existing commands, but don't add new commands.
> 
>     * Add new commands.
> 
> 
> Second and third points can reside in this PR as separate commits.
> 
> Feel free to discuss any point with me.

I agree with @Totktonada in regard to cleaning up the patch set in tarantool/tarantool repo.

As for tarantool/luarocks I think that it's already clean enough. I don't agree with the branch name `tarantool-2`, because it's meaningless.

**Comment by zwirec** [comment]
@rosik ok. What do you think about cherry-pick'ing current patch set to luarocks v3? We didn't succeed to do it clearly.

**Comment by Totktonada** [comment]
I propose another way to work with forked repos.

Maintain set of branches like `luarocks-3.1.1-tarantool`, `luarocks-3.1.2-tarantool` based on corresponding upstream tags with our patches on top. Set a last of those branches as default in our repo. Use `cherry-pick -x` to fill a new branch with the patchset in order to track origins of commits. If a commit need to be significantly reworked to be placed on top of a new branch, then it is okay to create it from schratch and mention the original commit is a commit message.

Both ways have the same power, but this one does not involve rebasing of long-term branches.

@rosik Are you have objections against this way?

**Comment by Totktonada** [comment]
@zwirec @rosik Where I can put comments about `luarocks-3.x` patches? Can you open a PR to `luarocks-3.1.1-tarantool` branch if there are no objections against the proposal above? I had created the branch on 3.1.1 tag. Or you can send the patches to tarantool-patches mailing list (please, leave me in To or CC).

The general comment about the patches that they lack of mention of its origins (commits upward luarocks-2 on which they are based) and so it is harder to find a related discussion (a mail, issue or PR). For the commit 'fix strange behaviour' I would at least mention a command where it is observed. "test: fix reporting failures on 'command' backend" has 'cherry picked from' note with orphan commit (there is no such commit in my clone). Whether fetching submodules will still work with rockspecs of 2nd version (see #2609)? Also there is forgotten commented print :)

I'll look in depth when there will be a place to comment.

**Comment by rosik** [comment]
> I propose another way to work with forked repos.
> 
> Maintain set of branches like `luarocks-3.1.1-tarantool`, `luarocks-3.1.2-tarantool` based on corresponding upstream tags with our patches on top. Set a last of those branches as default in our repo. Use `cherry-pick -x` to fill a new branch with the patchset in order to track origins of commits. If a commit need to be significantly reworked to be placed on top of a new branch, then it is okay to create it from schratch and mention the original commit is a commit message.
> 
> Both ways have the same power, but this one does not involve rebasing of long-term branches.
> 
> @rosik Are you have objections against this way?

I agree with your proposal.

How do we name a branch, if it's not based on a release tag, but on intermediate commit (like `3.1.1-2`)?
Can we name it `luarocks-3.1.1-2-tarantool`?

**Comment by Totktonada** [comment]
> How do we name a branch, if it's not based on a release tag, but on intermediate commit (like `3.1.1-2`)?
> Can we name it `luarocks-3.1.1-2-tarantool`?

I guess this need will be rare, but anyway we can use the scheme you proposed (`luarocks-3.1.1-2-tarantool`) and use git describe to determine a number of commits upward a last tag (2 in the case).

File: CMakeLists.txt [hunk 2]
```diff
@@ -233,8 +233,8 @@ install(DIRECTORY DESTINATION ${MODULE_LIBDIR})
 include(cmake/multilib.cmake)
 
 set(_datadirs)
-list(APPEND _datadirs "/usr/local/share") # LuaRocks
 list(APPEND _datadirs "${CMAKE_INSTALL_FULL_DATADIR}") # Install prefix
+list(APPEND _datadirs "/usr/local/share") # LuaRocks
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply CMakeLists.txt]: Was removed from the patchset.

File: extra/luarocks/CMakeLists.txt [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
 install(DIRECTORY ${PROJECT_SOURCE_DIR}/third_party/luarocks/src/luarocks
     DESTINATION ${MODULE_LUADIR})
-install(FILES ${CMAKE_CURRENT_BINARY_DIR}/site_config.lua
-    DESTINATION ${MODULE_LUADIR}/luarocks/
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/hardcoded.lua
+    DESTINATION ${MODULE_LUADIR}/luarocks/core
     PERMISSIONS
     OWNER_READ OWNER_WRITE OWNER_READ
-    GROUP_READ WORLD_READ)
+    GROUP_READ WORLD_READ)
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply extra/luarocks/CMakeLists.txt]: I'll do.

**Comment by Totktonada** [comment]
I made the following changes:

* Fixed minor typos here and there (say, `<argument>>` → `<argument>`).
* Changed wording in the commit messages a bit.
* Referenced #4052 in the first commit.
* Prefixed all commits with 'rocks:' prefix (we usually prefix all commits to ease git log reading).
* Added generated hardcoded.lua into .gitignore.
* Changed the submodule branch name to `luarocks-3.1.1-tarantool` as we agreed before (it only matters for `git submodule update --remote`, but anyway).
* Rebased at top of current master.
* Fixed the submodule on the current luarocks-3.x HEAD.

I'll update the PR again when we'll update `luarocks-3.1.1-tarantool` branch finally (see https://github.com/tarantool/luarocks/pull/7).

**Comment by Totktonada** [comment]
Pushed updated luarocks-3.x branch into luarocks-3.1.1-tarantool. Updated the submodule in this PR.

**Pull request merged**